src/builder.hpp

```diff
@@ -0,0 +1,75 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#pragma once
+
+#include "parser/specification.hpp"
+#include <manifest.hpp>
+#include <memory>
+#include <parameter.hpp>
+#include <rule.hpp>
+#include <ruleset.hpp>
+#include <ruleset_info.hpp>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ddwaf {
+
+class builder {
+public:
+    using ptr = std::shared_ptr<builder>;
+
+    builder(object_limits limits, ddwaf_object_free_fn free_fn, ddwaf::obfuscator event_obfuscator)
+        : limits_(limits), free_fn_(free_fn), event_obfuscator_(std::move(event_obfuscator))
+    {}
+
+    ~builder() = default;
+    builder(builder &&) = default;
+    builder(const builder &) = delete;
+    builder &operator=(builder &&) = default;
+    builder &operator=(const builder &) = delete;
+
+    std::shared_ptr<ruleset> build(parameter object, ruleset_info &info);
+
+protected:
+    enum class change_state : uint32_t {
+        none = 0,
+        rules = 1,
+        overrides = 2,
+        filters = 4,
+        data = 8
+    };
+
+    friend constexpr change_state operator|(change_state lhs, change_state rhs);
+    friend constexpr change_state operator&(change_state lhs, change_state rhs);
+
+    change_state load(parameter::map &root, ruleset_info &info);
+
+    std::shared_ptr<ruleset> build_helper(parameter::map &root, ruleset_info &info);
+
+    object_limits limits_;
+    ddwaf_object_free_fn free_fn_;
+    ddwaf::obfuscator event_obfuscator_;
+
+    manifest target_manifest_;
+    std::unordered_map<std::string, std::string> rule_data_ids_;
+
+    parser::rule_spec_container base_rules_;
+    parser::rule_data_container dynamic_processors_;
+    parser::override_spec_container overrides_;
+    parser::filter_spec_container exclusions_;
+
+    std::unordered_map<std::string_view, rule::ptr> final_rules_;
+    rule_tag_map rules_by_tags_;
+    std::unordered_set<manifest::target_type> targets_from_rules_;
+
+    std::unordered_map<std::string_view, exclusion::rule_filter::ptr> rule_filters_;
+    std::unordered_map<std::string_view, exclusion::input_filter::ptr> input_filters_;
```

Comment:
This granular use of shared_ptrs is a bit on the wasteful side. For instance, take the rules. While you frequently manage to avoid having to change the refcount (when iterating taking the shared_ptr by reference), if the point is that rules, filters etc can be shared between rulesets, then it's those collections that should be behind the shared_ptr, not the individual filters and rules.

Reply:
I'll do this for 1.8.1